src/compact.jl

```diff
@@ -103,30 +115,29 @@ macro compact(fex, kwexs...)
 
   # edit expressions
   vars = map(ex -> ex.args[1], kwexs)
-  assigns = map(ex -> Expr(:(=), ex.args...), kwexs)
-  @gensym self
-  pushfirst!(fex.args[1].args, self)
-  addprefix!(fex, self, vars)
+  fex = supportself(fex, vars)
 
   # assemble
   return esc(quote
     let
-      $(assigns...)
-      $CompactLayer($fex, $name, ($layer, $input, $block), $setup; $(vars...))
+      $CompactLayer($fex, $name, ($layer, $input, $block), $setup; $(kwexs...))
     end
   end)
 end
 
-function addprefix!(ex::Expr, self, vars)
-  for i = 1:length(ex.args)
-    if ex.args[i] in vars
-      ex.args[i] = :($self.$(ex.args[i]))
-    else
-      addprefix!(ex.args[i], self, vars)
+function supportself(fex::Expr, vars)
+  @gensym self
+  @gensym curried_fex
+  # To avoid having to manipulate fex's arguments and body explicitly, we form a curried function first
+  # that wraps the full fex expression, and then uncurry it programatically rather than syntactically.
+  let_exprs = map(var -> :($var = $self.$var), vars)
+  return quote
+    $curried_fex = ($self) -> let $(let_exprs...) 
+        $fex
     end
+    ($self, args...; kwargs...) -> begin $curried_fex($self)(args...; kwargs...) end
   end
```

Comment:
Not this PR, but is there any way this let block trick could be used with `LayerFactory(x -> abs.(x .* w); w = rand(3))`, without a macro? 

I think you can't just paste it into `@generated function (f::LayerFactory)(args...)` as it creates closures. But is there some more clever way?

Reply:
I also pondered this. It seems difficult, because the `x -> abs.(x .* w)` evaluated as ordinary code would close over `w`, and it seems hard to revert that.

Reply:
(even if we could somehow do it, by deconstructing and reconstructing the closure, I'd argue it's too magical to *not* be in a macro!)